File: FSharp/Step1/SftpActors.fs
[<AutoOpen>]
module SftpActors
    open System
    open Akka.FSharp
    open SftpClient

    type SftpCommand =
        | Connect
        | Disconnect

    let sftpActor (clientFactory : IClientFactory) (mailbox: Actor<_>) =

        let rec disconnected () = 
            actor {
                let! message = mailbox.Receive ()
                match message with
                | Connect -> 
                    let connection = clientFactory.CreateSftpClient()
                    connection.Connect()
                    return! connected (connection)
                | _ ->
                    cprintfn ConsoleColor.Yellow "Sftp: invalid operation in disconnected state: %A" message

                return! disconnected ()
            } 
        and connected (connection) = 
            actor {
                let! message = mailbox.Receive ()
                match message with
                | Disconnect -> 
                    connection.Disconnect()
                    connection.Dispose()
                    return! disconnected ()
                | _ ->
                    cprintfn ConsoleColor.Yellow "Sftp: invalid operation in connected state: %A" message

                return! connected (connection)
            } 

        disconnected ()

File: CSharp/Step1/SftpActors.cs
using System;
using Akka.Actor;

using Shared;

namespace SftpActors
{
	public class SftpActor : ReceiveActor
	{
		private IClientFactory _clientFactory;
		private ISftpClient _connection;

		public SftpActor(IClientFactory clientFactory)
		{
			_clientFactory = clientFactory;

			Disconnected();
		}

		public class Connect { }
		public class Disconnect { }

		private void Disconnected()
		{
			Receive<Connect>((cmd) =>
			{
				_connection = _clientFactory.CreateSftpClient();
				_connection.Connect();

				Become(Connected);
			});
		}

		private void Connected()
		{
			Receive<Disconnect>((cmd) =>
			{
				_connection.Disconnect();
				_connection.Dispose();

				Become(Disconnected);
			});
		}
	}
}
